theano/gpuarray/dnn.py

```diff
@@ -2732,6 +2732,191 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+
+    def c_headers(self):
+        return ['cudnn.h', 'cudnn_helper.h']
+
+    def c_header_dirs(self):
+        return [os.path.dirname(__file__), config.dnn.include_path]
+
+    def c_libraries(self):
+        return ['cudnn']
+
+    def c_lib_dirs(self):
+        return [config.dnn.library_path]
+
+    def do_constant_folding(self, node):
+        return False
+
+    def __init__(self, dimensions, precision="float32"):
+        COp.__init__(self, ["c_code/spatialtf_desc.c"], "APPLY_SPECIFIC(spatialtf_desc)")
+
+        # dimensions must have at least width and height
+        assert len(dimensions) >= 2
+
+        self.dimensions = tuple(dimensions)
+
+        # cuDNN supports only 2D transformations, therefore output tensor must
+        # not exceed 4 dimensions (width, height, num_feature_maps, num_images)
+        assert len(self.dimensions) <= 4
+
+        assert cudnn.cudnnDataType_t.has_alias(precision)
+        self.precision = precision
+
+    def make_node(self):
+        node = Apply(self, [],
+                     [CDataType("cudnnSpatialTransformerDescriptor_t",
+                                freefunc="cudnnDestroySpatialTransformerDescriptor")()])
+        # DebugMode cannot compare the values of CDataType variables, so by
+        # default it returns False all the time. To prevent DebugMode from
+        # complaining because of the MergeOptimizer, we make this variable
+        # always compare to True.
+        out = node.outputs[0]
+        out.tag.values_eq_approx = tensor.type.values_eq_approx_always_true
+        return node
+
+    # Number of images
+    nimages = property(lambda self: self.dimensions[0])
+    # Number of channels
+    nchannels = property(lambda self: self.dimensions[1])
+    # Grid height
+    height = property(lambda self: self.dimensions[2])
+    # Grid width
+    width = property(lambda self: self.dimensions[3])
+    # Number of dimensions in the output tensor
+    nb_dims = property(lambda self: len(self.dimensions))
+
+    def c_code_cache_version(self):
+        return (super(GpuDnnSpatialTfDesc, self).c_code_cache_version(), version())
+
+
+class GpuDnnGridGenerator(DnnBase):
+
+    """
+    This Op builds a spatial transformer grid generator for use in spatial transformer network
+    operations.
+    """
+
+    __props__ = ('precision',)
+    _cop_num_inputs = 3
+    _cop_num_outputs = 1
+
+    def __init__(self, precision):
+        DnnBase.__init__(self, ["c_code/spatialtf_grid.c"], "spatialtf_grid")
+
+        self.precision = precision
+
+    def dnn_context(self, node):
+        return node.outputs[0].type.context_name
+
+    def make_node(self, grid_dimensions, theta, desc):
+        context_name = infer_context_name(desc, theta)
+
+        grid_dimensions = as_tensor_variable(grid_dimensions)
+        theta = gpu_contiguous(as_gpuarray_variable(theta, context_name))
+
+        assert cudnn.cudnnDataType_t.has_alias(theta.dtype)
```

Comment:
From your C code, you expect theta to be float, half or double, but cudnnDataType may include `int8` or other types (e.g. in cuDNN V6). So here, check explicitely that theta.dtype is in ('float16', 'float32', 'float64').